pkg/controller/direct/cloudbuild/externalresource.go

```diff
@@ -19,56 +19,57 @@ package cloudbuild
-	segments := strings.Split(extResRef, "/projects/")
-	if len(segments) != 2 {
-		return nil, fmt.Errorf("externalReference should be <baseUrl>/projects/<project>/locations/<location>/workerPools/<workerPool>, got %s",
-			extResRef)
-	}
-	segments = strings.Split(segments[1], "/")
-	if len(segments) == 5 && segments[1] == "locations" && segments[3] == "workerPools" {
-		return &ResourceRef{
-			project:     segments[0],
-			location:    segments[2],
-			resourceID:  segments[4],
-			externalRef: baseResourceRef.Get(),
-		}, nil
-	}
-	return nil, fmt.Errorf("externalReference should be in the form of <baseUrl>/projects/<project>/locations/<location>/workerPools/<workerPool>, got %s",
-		extResRef)
+type CloudBuildWorkerPoolIDentity struct {
+	project    string
+	location   string
+	workerpool string
 }
 
-type ResourceRef struct {
-	resourceID  string
-	location    string
-	project     string
-	externalRef *string
+// Parent builds a CloudBuildWorkerPool parent of the format projects/<project>/locations/<location>
+func (c *CloudBuildWorkerPoolIDentity) Parent() string {
+	return fmt.Sprintf("projects/%s/locations/%s", c.project, c.location)
 }
 
-func (e *ResourceRef) GetExternalReference() *string {
-	return e.externalRef
+// FullyQualifiedName builds a CloudBuildWorkerPool resource of the format projects/<project>/locations/<location>/workerPools/<workerPool>
+func (c *CloudBuildWorkerPoolIDentity) FullyQualifiedName() string {
+	return fmt.Sprintf("projects/%s/locations/%s/workerPools/%s", c.project, c.location, c.workerpool)
 }
 
-func (e *ResourceRef) GetResourceID() string {
-	return e.resourceID
+// fromID builds a externalRef from a CloudBuildWorkerPoolIDentity
+func asExternalRef(id *CloudBuildWorkerPoolIDentity) *string {
+	e := fmt.Sprintf("%sprojects/%s/locations/%s/workerPools/%s", serviceDomain, id.project, id.location, id.workerpool)
+	return &e
 }
 
-func (e *ResourceRef) GetLocation() string {
-	return e.location
+// asID builds a CloudBuildWorkerPoolIDentity from a externalRef
+func asID(externalRef string) (*CloudBuildWorkerPoolIDentity, error) {
+	if !strings.HasPrefix(externalRef, serviceDomain) {
+		return nil, fmt.Errorf("externalRef shall has prefix %s, got %s", serviceDomain, externalRef)
+	}
+	path := strings.TrimPrefix(externalRef, serviceDomain)
+	tokens := strings.Split(path, "/")
+	if len(tokens) != 6 || tokens[0] != "projects" || tokens[2] != "locations" || tokens[4] != "workerPools" {
+		return nil, fmt.Errorf("externalRef should be %s/projects/<project>/locations/<location>/workerPools/<workerPool>, got %s",
+			serviceDomain, externalRef)
+	}
+	return &CloudBuildWorkerPoolIDentity{
+		project:    tokens[1],
+		location:   tokens[3],
+		workerpool: tokens[5],
+	}, nil
 }
 
-func (e *ResourceRef) GetProject() string {
-	return e.project
+// fromRaw builds a CloudBuildWorkerPoolIDentity from resource components.
+func fromRaw(project, location, workerpool string) *CloudBuildWorkerPoolIDentity {
```

Comment:
Nit: fromRaw is not the most obvious name when you're calling it.  Maybe `buildCloudBuildWorkerPoolIdentity` ?  (and then the string-parsing version could be `parseCloudBuildWorkerPoolIdentity`)

Because this is in the CloudBuild package we could also do `buildWorkerPoolIdentity`

Reply:
So I defined 3 methods here and want them to discovered with a clear indication of their usage:
- `asID` asks for string ID and returns structured ID
- `asExternalRef` asks for structured ID and returns the string ID. It is the opposite of `asID`
- `fromRaw` asks for the each segment value, and returns the structured ID. 
-